### PR TITLE
Hw3

### PR DESCRIPTION
**Pull request by vbelymenko**

### CODE REVIEW

File: src/components/course-list/CourseList.jsx [hunk 1]
```diff
@@ -1,43 +1,34 @@
...
-
-    handleEditCourseClick = (id) => {
-        this.props.history.push(`courses/details/${id}`);
-    }
-
-    handleRemoveCourseClick = (id) => {
-        remove(id);
-    }
-
     renderCourses = () => {
```
Comment (denys-kostenko) [inline src/components/course-list/CourseList.jsx:26]: components own methods under `render()`

File: src/components/edit-course/EditCourse.jsx [hunk 2]
```diff
@@ -12,17 +12,18 @@ export class EditCourse extends Component {
...
                 date: PropTypes.string,
-                duration: PropTypes.string
+                duration: PropTypes.number
             }).isRequired,
             courseAuthors: PropTypes.arrayOf(PropTypes.number).isRequired,
             possibleAuthors: PropTypes.arrayOf(PropTypes.number).isRequired,
             onChange: PropTypes.func.isRequired,
-            onSave: PropTypes.func.isRequired
+            onSave: PropTypes.func.isRequired,
+            onCancle: PropTypes.func.isRequired
```
Comment (denys-kostenko) [inline src/components/edit-course/EditCourse.jsx]: `onCancel` not `onCancle`

File: src/components/edit-course/EditCourse.jsx [hunk 3]
```diff
@@ -110,7 +111,7 @@ export class EditCourse extends Component {
                     <button onClick={onSave} className="btn btn-info btn-lg">
                         Save
                     </button>
-                    <button className="btn btn-danger btn-lg">
+                    <button onClick={onCancle} className="btn btn-danger btn-lg">
                         Cancle
```
Comment (denys-kostenko) [inline src/components/edit-course/EditCourse.jsx:115]: again Cancel...

File: src/components/navigation/Navigation.jsx [hunk 4]
```diff
@@ -4,25 +4,44 @@ import TextField from '@material-ui/core/TextField';
 import Search from '@material-ui/icons/Search';
 import Add from '@material-ui/icons/Add';
 import './Navigation.css';
+import PropTypes from 'prop-types';
 
 export class Navigation extends React.Component {
 
+    constructor(props) {
```
Comment (denys-kostenko) [inline src/components/navigation/Navigation.jsx]: Again https://github.com/vbelymenko/react-app/pull/2#discussion_r235110441

File: src/components/navigation/Navigation.jsx [hunk 5]
```diff
@@ -4,25 +4,44 @@ import TextField from '@material-ui/core/TextField';
...
+        super(props);
+        this.state = { filter: '' };
+    }
+
+    static get propTypes() {
+        return {
+            onAdd: PropTypes.func.isRequired,
```
Comment (denys-kostenko) [inline src/components/navigation/Navigation.jsx]: what does another dev can understand from prop `onAdd` or `onFilter`???
what add? etc. `addCourse` more better and meaningful

Please think about naming and give good names to actions and thunks. Not just in that component

File: src/page/CoursesPage.jsx [hunk 6]
```diff
@@ -2,15 +2,54 @@ import React, { Component } from "react";
 import { AppContainer } from "../components/app-container";
 import { CourseList } from "../components/course-list";
 import { Navigation } from "../components/navigation";
+import { withRouter } from "react-router-dom";
+import { connect } from 'react-redux';
+import { remove, getAll, updateFilter, cleanFilter } from '../store/actions/courses';
 
-export class CoursesPage extends Component {
+class CoursesContainer extends Component {
+
```
Comment (denys-kostenko) [inline src/page/CoursesPage.jsx:11]: Where PropTypes???

File: src/page/CoursesPage.jsx [hunk 7]
```diff
@@ -2,15 +2,54 @@ import React, { Component } from "react";
...
+    componentDidMount() {
+        this.props.getAll();
+    }
+
+    componentWillUnmount(){
+        this.props.cleanFilter();
+    }
+
+    handleEditCourseClick = (id) => {
```
Comment (denys-kostenko) [inline src/page/CoursesPage.jsx]: again about functions order inside component https://github.com/vbelymenko/react-app/pull/2#discussion_r235110441

File: src/page/CoursesPage.jsx [hunk 8]
```diff
@@ -2,15 +2,54 @@ import React, { Component } from "react";
...
+        this.props.history.push(`courses/details/${id}`);
+    }
+
+    handleAddCourseClick = () => {
```
Comment (denys-kostenko) [inline src/page/CoursesPage.jsx]: `handleAddCourseClick` just `handleAddCourse` etc.

File: src/page/CoursesPage.jsx [hunk 9]
```diff
@@ -2,15 +2,54 @@ import React, { Component } from "react";
...
             </AppContainer>
         );
     }
-}
\ No newline at end of file
+}
+
+const mapStateToProps = (state) => {
+    return {
+        courses: state.courses.coursesList.filter(course=> course.title.includes(state.courses.filter))
```
Comment (denys-kostenko) [inline src/page/CoursesPage.jsx]: move it to selector. like SelectCourses = (state) => state.courses.........

File: src/page/CoursesPage.jsx [hunk 10]
```diff
@@ -2,15 +2,54 @@ import React, { Component } from "react";
...
+    };
+}
+
+export const CoursesPage = withRouter(connect(mapStateToProps, { getAll, remove, updateFilter, cleanFilter })(CoursesContainer));
```
Comment (denys-kostenko) [inline src/page/CoursesPage.jsx]: better to write each object key from new line, more readdable:
```
export const CoursesPage = withRouter(connect(mapStateToProps, { 
getAll, 
remove, 
updateFilter, 
cleanFilter 
})(CoursesContainer));

```

Comment (denys-kostenko) [reply src/page/CoursesPage.jsx]: again about actions names, what remove? etc

Comment (denys-kostenko) [reply src/page/CoursesPage.jsx]: Also  you can wrap you page in hoc App, for example, and wrap in with router just App

File: src/page/CreateCoursePage.jsx [hunk 11]
```diff
@@ -1,58 +1,52 @@
...
             </AppContainer>
         );
     }
 }
 
-export const CreateCoursePage = withRouter(CreateCoursePageContainer);
\ No newline at end of file
+const mapStateToProps = (state) => {
+    return {
+        course: state.course.courseEntity
```
Comment (denys-kostenko) [inline src/page/CreateCoursePage.jsx]: move it to selector

File: src/page/EditCoursePage.jsx [hunk 12]
```diff
@@ -1,55 +1,52 @@
...
             </AppContainer>
         );
     }
 }
 
-export const EditCoursePage = withRouter(EditCoursePageContainer);
\ No newline at end of file
+const mapStateToProps = (state) => {
+    return {
+        course: state.course.courseEntity
```
Comment (denys-kostenko) [inline src/page/EditCoursePage.jsx]: again i see `state.course.courseEntity` that why we should keep that logic in selectors near it's store

File: src/store/actions/courses.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
+import * as api from '../api/CoursesApi';
+import * as courseAction from './courseActions';
+
```
Comment (denys-kostenko) [inline src/store/actions/courses.js]: шо за простыня? Мы ж на встрече говорили за то как стор должен выглядеть, структура папок должна стремится к тому как выглядет json

File: src/store/api/CoursesApi.js [hunk 14]
```diff
@@ -11,4 +11,8 @@ export const remove = (id) => apify(axios.delete(`${apiUrl}/${id}`));
```
Comment (denys-kostenko) [inline src/store/api/CoursesApi.js]: Api that separate app logic, move it on level upper

File: src/store/reducers/course.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import * as types from '../actions/actionTypes';
```
Comment (denys-kostenko) [inline src/store/reducers/course.js]: That how store should looks like 
image

File: src/components/edit-course/EditCourse.jsx (hunk 2 above)
Comment (vbelymenko) [reply src/components/edit-course/EditCourse.jsx]: fixed

File: src/components/edit-course/EditCourse.jsx (hunk 3 above)
Comment (vbelymenko) [reply src/components/edit-course/EditCourse.jsx:115]: fixed

File: src/components/course-list/CourseList.jsx (hunk 1 above)
Comment (vbelymenko) [reply src/components/course-list/CourseList.jsx:26]: fixed

File: src/components/navigation/Navigation.jsx (hunk 5 above)
Comment (vbelymenko) [reply src/components/navigation/Navigation.jsx]: fixed

File: src/page/CoursesPage.jsx (hunk 7 above)
Comment (vbelymenko) [reply src/page/CoursesPage.jsx]: fixed

File: src/page/CoursesPage.jsx (hunk 8 above)
Comment (vbelymenko) [reply src/page/CoursesPage.jsx]: fixed

**Comment by vbelymenko** [comment]
@denys-kostenko , i pushed PR where i fixed some isues. I marked as "fixed" in conversasion which issues were fixed. Another issues will be fixed tommorow. I will notify you. 
Thanks

File: src/store/reducers/course.js (hunk 15 above)
Comment (vbelymenko) [reply src/store/reducers/course.js]: fixed

File: src/store/actions/courses.js (hunk 13 above)
Comment (vbelymenko) [reply src/store/actions/courses.js]: fixed

File: src/page/CreateCoursePage.jsx (hunk 11 above)
Comment (vbelymenko) [reply src/page/CreateCoursePage.jsx]: fixed

File: src/page/EditCoursePage.jsx (hunk 12 above)
Comment (vbelymenko) [reply src/page/EditCoursePage.jsx]: fixed

File: src/page/CoursesPage.jsx (hunk 9 above)
Comment (vbelymenko) [reply src/page/CoursesPage.jsx]: fixed

**Pull request merged**